client/state/http/index.js

```diff
@@ -67,22 +63,24 @@ export const httpHandler = ( { dispatch }, action ) => {
 		)
 		.join( '&' );
 
-	if ( queryString.length > 0 ) {
-		request.query( queryString );
-	}
-
-	headers.forEach( ( [ headerKey, headerValue ] ) => request.set( headerKey, headerValue ) );
+	try {
+		const response = await fetch( queryString.length ? `${ url }?${ queryString }` : url, {
+			method,
+			headers: fetchHeaders,
+			body,
+			credentials: withCredentials ? 'include' : 'same-origin',
+		} );
 
-	request.accept( 'application/json' );
-
-	if ( body ) {
-		request.send( body );
+		if ( response.ok ) {
+			dispatch( extendAction( onSuccess, successMeta( { body: await response.json() } ) ) );
+		} else {
+			dispatch(
+				extendAction( onFailure, failureMeta( { response: { body: await response.json() } } ) )
+			);
+		}
+	} catch ( error ) {
+		dispatch( extendAction( onFailure, failureMeta( error ) ) );
```

Comment:
the semantics are slightly different here from the original. Now if the dispatch of onSuccess throws, onFailure will also be thrown. Prior, only onSuccess would be dispatched.

Reply:
Hmm, good point; I didn't consider the possibility of `dispatch` throwing.

I moved the fetching and JSON parsing into its own block, together with that error catching.